daffodil-cli/src/test/scala/org/apache/daffodil/CLI/Util.scala

```diff
@@ -241,33 +241,101 @@ object Util {
     val expect = eb.build()
     val tester = new CLITester(expect, toIn)
 
-    try {
-      testFunc(tester)
-    } finally {
-      threadOrProc match {
-        case Left(thread) => thread.join(timeout * 1000)
-        case Right(process) => process.waitFor(timeout, TimeUnit.SECONDS)
-      }
-      expect.close()
-      toIn.close()
-      fromOut.close()
-      fromErr.close()
+    // run the test function in a new thread. We do this so that we can configure
+    // ExpectIt to have an infinite timeout. This way we can trigger a timeout only
+    // after the has CLI exited, at which point we give the test thread timeout
+    // seconds to finish its tests and interrupt it if it still hasn't finished. This
+    // way if the CLI is slow to start up, ExpectIt won't create a timeout exception
+    // and fail the test. We at least let the CLI finish before doing that
+    val testThread = new TestThread(testFunc, tester, threadOrProc)
+    testThread.start()
+
+    // wait for the CLI process/thread to finish with no timeout--we really
+    // don't know how long the CLI is going to take, but we assume the CLI will
+    // eventually finish. If it doesn't, we'll see the hang and investigate
+    threadOrProc match {
+      case Left(thread) => thread.join()
+      case Right(process) => process.waitFor()
     }
 
+    // now that the CLI has finished, we give the test thread timeout seconds
+    // to finish verifying the output. Hopefully it was doing all the checks in
+    // parallel with the CLI thread so it should be mostly complete and not
+    // need the full timeout
+    testThread.join(timeout * 1000)
+
+    // If the test thread still isn't done, assume it failed and is stuck expecting
+    // something that isn't coming. Interrupt the thread to end it. If it is already
+    // finished, this does nothing.
+    testThread.interrupt()
+
+    // at this point we may need to wait for the test thread to die from the
+    // interrupt. Expect might be doing some cleanup as it handles the interruption
+    while (testThread.isAlive) Thread.sleep(100)
```

Comment:
In your pull request, you say:

> If a particular code hunk should never time out [...]. This means if the test ever fails we might hang forever, but that is unlikely, and it's better than getting random false negatives and trying to determine if they are spurious or not

Is this line the place you were talking about where we might hang forever?  If so, please include the second quoted sentence in the comment preceding this line so that readers will know we already considered the possibility of the while loop never terminating and are okay with it.

Reply:
I don't like these unbounded loops. 

I think if this spins 10 times (1 full second) or maybe 20 times, then it should start outputting something to somewhere like stderr indicating that it seems to be stuck.

Reply:
> Is this line the place you were talking about where we might hang forever?

This is one potential place. Really the only way for this loop to hang is if the expect library ignores the above interrupt (e.g. catches an `InterruptedException`, ignores it, and continues doing expect stuff). In my development of this stuff, it seems the expect library handles interrupts as expected. And I do see logic in the expect code to make sure it fails expect()s if it has been interrupted.

And now that I think about it, I would actually prefer to change this line to `testThread.join()` to wait until the test thread dies from the interrupt. I think `join` is just implemented as a very similar `isAlive` loop, so this would be cleaner.

Another place things could potentially hang with these changes is in this code above:
```scala
   threadOrProc match {
      case Left(thread) => thread.join()
      case Right(process) => process.waitFor()
    }
```
So if the CLI gets stuck in an infinite loop or is blocked waiting for input then this will hang forever. There really isn't much looping logic in the CLI itself, so if this did hang it probably means there is a bug Daffodil core causing an infinite loop, which would probably also cause TDML tests to hang. Or more likely it means the test body is broken and the CLI is expecting input on stdin and the test body isn't providing it.

The other places that could now hang are the tests related to socket reading where there are potentially blocking reads. But those are very unlikely to hang since the tests are written correctly so the reads shouldn't block, and Java isn't likely to change the blocking behavior of sockets, so our tests making sure reads don't block will likely never fail and never need to change.

I do agree that it's worth adding comments to these places in the CLI where a hang is theoretically possible and what could cause it.

> I think if this spins 10 times (1 full second) or maybe 20 times, then it should start outputting something to somewhere like stderr indicating that it seems to be stuck.

Just playing devils advocate, we don't do anything like this for any other tests, either unit or TDML. And IMO, all our other tests have just as much a likelihood to to enter an infinite loop as these tests (which I argue are both very unlikely). And if they happened, then it would be consistent and easily repeatable.

I do admit that these CLI tests are definitely more complicated with the different thread interactions, but expect ignoring interrupts feels very unlikely. So practically the only place this runCLI will hang is with a infinite loop in the CLI (also feels unlikely, and would likely show up in TDML tests that would hang) or if the test body is broken and leaves the CLI hanging (e.g. the CLI is at a debugger breakpoint waiting for debugger commands, the CLI is reading from stdin and doesn't provide the right input/never closes stdin).

In fact, we had one test that actually did leave the CLI hanging waiting for debug input and these changes made it very obvious and repeatable. The test hung at the cliThread join() call, it needed a Ctrl-C to kill the test, and sbt output the exception that looked like this

```
[error] Test org.apache.daffodil.cliTest.TestCLIdebugger.test_3585_CLI_Debugger_simpleDebugger_unparse failed: java.lang.InterruptedException: null, took 12.068 sec
[error]     at java.lang.Object.wait(Native Method)
[error]     at java.lang.Thread.join(Thread.java:1257)
[error]     at java.lang.Thread.join(Thread.java:1331)
[error]     at org.apache.daffodil.CLI.Util$.runCLI(Util.scala:257)
[error]     at org.apache.daffodil.cliTest.TestCLIdebugger.test_3585_CLI_Debugger_simpleDebugger_unparse(TestCLIDebugger.scala:873)
```

So I knew exactly which test it was and which join it was hanging on. And it turned out the test was just broken and missing a debugger `"quit"` command, fixed in this PR.

We could definitely put a time limit on the CLI joins and start outputting messages (like "CLI loop is taking longer than expected") like you suggest, but those are just going to be false negatives more often than not. Tests taking longer than expected due to resource constrains is what led to randomly failing PRs. Random error messages with tests still passing may just confuse things more. And with a Ctrl-C we get precise information about which test and which join was hung.

Reply:
All that to say, this would be a reasonable approach and I don't think is too messy if we want a warning for the most likely case of a bad test causing the CLI to hang:

```scala
    val cliWarningTimeoutMS = 4 * timeout * 1000
    var isAlive = true
    while (isAlive) {
      isAlive = cliThreadOrProc match {
        case Left(cliThread) => {
          cliThread.join(cliWarningTimeoutMS)
          cliThread.isAlive
        }
        case Right(cliProcess) => {
          cliProcess.waitFor(cliWarningTimeoutMS, TimeUnit.MILLISECONDS)
          cliProcess.isAlive
        }
      }
      if (isAlive) {
        System.err.println("CLI test taking significantly longer than expeced--may be hung waiting on debugger input or closed stdin")
      }
    }
```

So every 4x timeout we output a message if the CLI hasn't completed yet. I'm not sure if 4x the timeout it too long or not long enough, but I really want to avoid false negatives.

I'm still not sure we need this extra complication for the `threadTest.join()` though--the likelihood of that hanging after an interrupt feels very low to me. A bad test causing the CLI to hang does feel much more possible though.

Reply:
Are you saying you could replace line 240,

`while (testThread.isAlive) Thread.sleep(100)`

with this much longer piece of code?  That is, replace Thread.sleep (putting the current thread to sleep until the test thread dies) with a cliThread.join or cliProcess.join (waiting for the CLI thread/process to die)?  These are different threads, so I think you meant another place for your new piece of code?

Reply:
Sorry, that comment got a longer than I expected and I may have made things more confusing.

I suggest changing line 240
```scala
while (testThread.isAlive) Thread.sleep(100)
``` 
With this:
```scala
testThread.join()
```
This is functionally the same thing but a bit cleaner. Neither approach should ever hang in practice because the expect library handles interrupts correctly and will die with the `interrupt()` call right above this line.

For the CLI join/waitFor, which is more likely to hang but mostly requires a broken test, if we want a message I suggest we replace this:
```scala
threadOrProc match {
  case Left(thread) => thread.join()
  case Right(process) => process.waitFor()
}
```

with this:

```scala
val cliWarningTimeoutMS = 4 * timeout * 1000
var isAlive = true
while (isAlive)
  isAlive = cliThreadOrProc match {
    case Left(cliThread) => {
      cliThread.join(cliWarningTimeoutMS)
      cliThread.isAlive
    }
    case Right(cliProcess) => {
      cliProcess.waitFor(cliWarningTimeoutMS, TimeUnit.MILLISECONDS)
      cliProcess.isAlive
    }
  }
  if (isAlive) {
    System.err.println("CLI test taking significantly longer than expected--" +
      "may be hung waiting on debugger input or closed stdin")
  }
}
```

This will still hang like before, but would output a message to stderr every (4 * timeout) seconds that the CLI doesn't end end.

If that's too messy we could probably make some helper functions that do something similar.

Reply:
OK, I'll defer to Mike since that was his request.  I'll say this, however, the phrase "---may be hung waiting on debugger input or closed stdin" is still confusing (at least to me) since it assumes some context the reader should know.  I'm not sure how a bad test can cause "waiting on debugger input or closed stdin"; should we just say "---check whether CLI test is broken" or something like that?

Reply:
The likely case isn't that the CLI is broken, but that a CLI test is broken. For example, we could have a CLI test that does this:


```scala
    runCLI(args"--debug parse ...") { cli =>
      cli.expect("(debug)")
      cli.sendLine("break foo")
      cli.sendLine("continue")
      cli.expect("(debug")
    } (ExitCode.Success)
```

So we start the CLI debugger, set a breakpoint, and continue until that breakpoint is hit. At this point the CLI is paused waiting for input, but the test doesn't provide anymore input. The CLI is now hung because no more input it coming from the test body, but the CLI doesn't know that--it can't differentiate between "no more input is coming" and "we are limited in resources and the input is just coming really really slowly".

So this isn't a bug in the CLI, it's a bug in the test not giving the CLI enough information to finish.

The other example I can think of is if daffodil schema needs to hit EOF to know when to finish parsing/unparsing, but a CLI test doesn't close stdin. So the CLI thinks more data is coming because there's no EOF, but the test body never sends anymore data or closes stdin.

I think these are the most common cases of a broken tests, so I tried to describe them (very tersely) in the error message. But I'm fine with a more generic "broken message" if the message is more confusing than helpful.

Reply:
OK, the reader needs to read that much detailed information to understand the message better.  Since we print that message every 4  * timeout seconds, the message needs to be terse, however.  The message will have to tell the reader the CLI test may be hanging and a place where to read the more detailed explanation you just wrote.  Something like:

```
CLI test may be hanging--see DAFFODIL-2751
```

Then you can add the detailed explanation you wrote to the issue's description so that anyone can understand how a bad CLI test might hang the CI workflow.  The description should say in conclusion, please debug the test to rule out these possible causes rather than "we are limited in resources and the input is just coming really slowly".  If it would be better, you can create a new issue just to serve as the place to read the explanation.